# Análisis de datos con Pyhon - Pre-entrega
Nombre y apellido: CESAR MORALES

# Etapa 1

## Script básico
Actividades 2 y 3

In [2]:
# Generamos una lista de diccionarios que contenga algunos datos
ventas = [
  {'producto': 'notebook', 'precio': 1500.0, 'cantidad': 3},
  {'producto': 'monitor', 'precio': 500.0, 'cantidad': 5},
  {'producto': 'mouse', 'precio': 20.0, 'cantidad': 15}
]

In [ ]:
# Calculamos las ventas acumuladas - map


In [4]:
# Calculamos las ventas acumuladas - listas por comprensión
Ventas_Acumuladas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])

## Import Datasets
Actividad 1

In [5]:
import pandas as pd

In [6]:
# Import dataset ventas
df_ventas = pd.read_csv("https://raw.githubusercontent.com/moralesmcesare/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/ventas.csv")

In [7]:
# Import dataset clientes
df_clientes = pd.read_csv("https://raw.githubusercontent.com/moralesmcesare/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/clientes.csv")

In [8]:
# Import dataset marketing
df_marketing = pd.read_csv("https://raw.githubusercontent.com/moralesmcesare/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/marketing.csv")

In [9]:
df_ventas

,id_venta,producto,precio,cantidad,fecha_venta,categoria
0,792,Cuadro decorativo,$69.94,5.0,02/01/2024,Decoración
1,811,Lámpara de mesa,$105.10,5.0,02/01/2024,Decoración
2,1156,Secadora,$97.96,3.0,02/01/2024,Electrodomésticos
3,1372,Heladera,$114.35,8.0,02/01/2024,Electrodomésticos
4,1546,Secadora,$106.21,4.0,02/01/2024,Electrodomésticos
...,...,...,...,...,...,...
3030,1837,Horno eléctrico,$104.12,9.0,30/12/2024,Electrodomésticos
3031,2276,Laptop,$85.27,9.0,30/12/2024,Electrónica
3032,2696,Laptop,$107.81,4.0,30/12/2024,Electrónica
3033,2913,Smartphone,$99.85,7.0,30/12/2024,Electrónica


## Dataframe ventas

### Análisis exploratorio de **datos**
Actividad 4

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [10]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [12]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head()

,id_venta,producto,precio,cantidad,fecha_venta,categoria
0,792,Cuadro decorativo,$69.94,5.0,02/01/2024,Decoración
1,811,Lámpara de mesa,$105.10,5.0,02/01/2024,Decoración
2,1156,Secadora,$97.96,3.0,02/01/2024,Electrodomésticos
3,1372,Heladera,$114.35,8.0,02/01/2024,Electrodomésticos
4,1546,Secadora,$106.21,4.0,02/01/2024,Electrodomésticos


In [11]:
# Usar sample() para tomar registros al azar
df_ventas.sample(5)

,id_venta,producto,precio,cantidad,fecha_venta,categoria
664,1164,Aspiradora,$102.32,2.0,17/03/2024,Electrodomésticos
1014,1203,Microondas,$96.53,1.0,22/04/2024,Electrodomésticos
2508,2285,Auriculares,$57.38,1.0,26/10/2024,Electrónica
2801,936,Cortinas,$110.96,2.0,01/12/2024,Decoración
2001,747,Elementos de cerámica,$32.85,10.0,22/08/2024,Decoración


In [13]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].sample(5)

,precio
175,$27.61
343,$97.01
2752,$55.58
263,$101.96
949,$45.79


In [14]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(5)

,cantidad
496,1.0
2563,3.0
1648,10.0
1235,8.0
1297,6.0


In [15]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"].sample(5)

,fecha_venta
2068,30/08/2024
2979,23/12/2024
2881,12/12/2024
24,05/01/2024
2412,12/10/2024


In [16]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

,count
categoria,
Decoración,1015
Electrónica,1012
Electrodomésticos,1008


### Calidad de datos
Actividad 5

In [23]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [31]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="producto")
# df_ventas[df_ventas.duplicated(keep="first")]

,id_venta,producto,precio,cantidad,fecha_venta,categoria
845,300,Adorno de pared,$67.75,9.0,06/04/2024,Decoración
833,300,Adorno de pared,$67.75,9.0,06/04/2024,Decoración
857,345,Alfombra,$56.78,9.0,07/04/2024,Decoración
832,145,Alfombra,$41.72,8.0,06/04/2024,Decoración
868,345,Alfombra,$56.78,9.0,07/04/2024,Decoración
...,...,...,...,...,...,...
854,2108,SmartWatch,$79.37,10.0,06/04/2024,Electrónica
831,2778,SmartWatch,$70.66,7.0,05/04/2024,Electrónica
825,2778,SmartWatch,$70.66,7.0,05/04/2024,Electrónica
877,2373,Smartphone,$123.60,12.0,07/04/2024,Electrónica


In [21]:
# Buscamos datos nulos
# df_ventas[df_ventas.isnull().any(axis=1)]
df_ventas.isnull().sum()

,0
id_venta,0
producto,0
precio,2
cantidad,2
fecha_venta,0
categoria,0


In [20]:
# Visualizar los nulos
df_ventas[df_ventas["precio"].isnull()]


,id_venta,producto,precio,cantidad,fecha_venta,categoria
136,627,Elementos de cerámica,NaN,NaN,17/01/2024,Decoración
139,2171,Parlantes Bluetooth,NaN,NaN,17/01/2024,Electrónica


In [27]:
# Analizar los precios de los NaN a ver que opciones tenemos
# df_ventas[df_ventas["producto"]== "Elementos de cerámica"]
# df_ventas["producto"].unique()
df_ventas["producto"].value_counts()

,count
producto,
Lámpara de mesa,181
Auriculares,147
Microondas,135
Cafetera,118
Laptop,104
Secadora,103
Jarrón decorativo,102
Smartphone,102
Alfombra,102


### Observaciones




* cantidad de registros: 3035
* cantidad de columnas: 6
* datos nulos: 2 1% del faltante y voy a poner dropna
* campo precio: encontré signo '$' habría que eliminarlo y luego cambiar dtype a float64
* campo fecha: ajustar a dtype datetime

## Dataframe marketing

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [ ]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

In [ ]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_marketing.head()

In [ ]:
# Analizamos el campo costo
df_marketing["costo"].sample(5)

In [ ]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(5)

In [ ]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(5)

### Calidad de los datos


In [ ]:
# Buscamos registros duplicados


### Observaciones
*** Completar ****

## Dataframe clientes

### Analisis exploratorio de datos

### Observaciones

# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas

* 35 duplicados eliminados - ok - se eliminaron 35 duplicados
* datos nulos: 2 - 1% del faltante dropna - se eliminaron 2 registros con 2 NaN
* campo precio: cambiar a float64, encontre signo '$' y luego cambiar dtype a float64
* campo fecha: ajustar a dtype datetime
* se observan 35 duplicados

In [32]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [33]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

np.int64(0)

In [36]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [59]:
# Validamos que no haya nulos
# df_ventas.isnull().sum()
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2998 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     2998 non-null   int64         
 1   producto     2998 non-null   object        
 2   precio       2998 non-null   int64         
 3   cantidad     2998 non-null   float64       
 4   fecha_venta  2998 non-null   datetime64[ns]
 5   categoria    2998 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 164.0+ KB


In [67]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].astype(str).str.strip().str.replace("$","")

In [61]:
# Validamos que precio no contenga $


In [48]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype("float")

In [49]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["precio"] = df_ventas["precio"].astype("int64")


In [50]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes

dtype('float64')

In [66]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [55]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

dtype('<M8[ns]')

### Dataframe marketing

*** Acciones ***

In [ ]:
# Convertir el campo fecha_inicio a dtype datetime


In [ ]:
# Convertir el campo fecha_fin a dtype datetime


In [ ]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas


In [ ]:
# Visualizamos


In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo


In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante


In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)


In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")


## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
# Agrupamos por categoria y agregamos la suma


In [ ]:
# Renombrar nombre de la columna


In [ ]:
# Ordenar y mostrar los resultados


## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'


In [ ]:
# Verificar coincidencias (muestra)


### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"


In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"


In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
